# CH9 COMPLEMENTS D'ALGORITHMIQUE

## 1. Concevoir et tester ses programmes

<h3> a. Les jeux de tests </h3>

Il arrive très souvent (même aux meilleurs ! ) de faire des erreurs en écrivant un code. Pour détecter ces éventuelles erreurs, on peut utiliser le programme (ou partie de programme) sur quelques cas concrets et vérifier qu'il produit effectivement les résultats attendus. 

On souhaite par exemple coder et tester une fonction nommée indice_maximum_tableau qui renvoie l'indice du plus grand élément de la liste. On peut la tester et vérifier que **indice_maximal_tableau([2 , 3 , 1])** vaut bien 1. Si le résultat est erroné, on sait qu'il y a une erreur dans le code, si le résultat est juste, cela ne signifie pas que tout est correct pour autant. Il faut donc essayer de créer un "jeu de tests" avec plusieurs tests les plus pertinents possibles (valeurs limites, cas particuliers ...).

Plutôt que d'effectuer les tests "à la main", on peut les inclure dans le programme, grâce à l'instruction **assert** et un test d'égalité. 

In [16]:
def indice_maximum_tableau(tableau):
    """ Renvoie l'indice du maximum d'un tableau (tableau) ou None s'il est vide"""
    
    max = tableau[0]
    for i in range(1, len(tableau)):
        if tableau[i] > max:
            max = tableau[i]
            indice_max = i
    return indice_max

assert indice_maximum_tableau([2 , 3 , 1]) == 1, "non, l'indice max n'est pas 1"
assert indice_maximum_tableau([2 , 3 , 1]) == 0, "non plus, l'indice max n'est pas 0"

AssertionError: non plus, l'indice max n'est pas 0

Si l'un des tests échoue, il faut bien sur relancer tous les tests.
Ajoutez quelques tests au programme précédent et vérifier que la fonction effectue bien ce qui est demandé (que se passe-t-il notamment si plusieurs éléments correspondent au maximum ? si le tableau est vide ? ). Modifier la fonction en conséquence.

<i> <b> Exercices d'application </b> </i>

On prétend que les fonctions suivantes testent l'appartenance de la valeur v au tableau t. 

1) Donner des tests pour ces fonctions, et en particulier un test montrant qu'elle sont incorrectes.

In [8]:
def appartient1(v, t):
    i = 0
    while i < len(t)-1 and t[i] != v:
        i = i + 1
    return i < len(t)

#assert appartient1(1,[0,2,5])==False

def appartient1(v, t):
    i = 0
    while i < len(t)-1 and t[i] != v:
        i = i + 1
    return i == len(t)

assert appartient1(1,[0,2,5])==False

In [10]:
def appartient2(v, t):
    for i in range(len(t)):
        if t[i] == v:
            trouvee = True
        else:
            trouvee = False
    return trouvee

assert appartient1(1,[0,1,5])==True

AssertionError: 

2) A partir de la fonction ci-dessous, on souhaite vérifier que l'utilisation se fait avec les bons paramètres, c'est à dire que t est bien une liste et que v est bien un entier. Intégrer des jeux de tests cette fois dans la fonction.

In [26]:
def appartient3(v, t):
    """ Fct correcte 
        Renvoie True si v appartient à t, False sinon """
    i = 0
    assert (type(t)==list)==True, "t doit être une liste"
    assert (type(v)==int)==True, "v doit être un entier"

    trouvee = False
    while i < len(t)-1:
        if t[i] == v:
            i = len(t)
            trouvee = True
        else:
            i = i + 1
    return trouvee

appartient3(1.2,[0,1,5])

AssertionError: v doit être un entier

<h3> b. Validité d'un algorithme : invariant de boucle et terminaison </h3>

Quand un programme contient une boucle, il convient de se poser plusieurs questions : les variables sont-elles correctement initialisées avant la boucle ? Le nombre de tours de la boucle est-il le bon ? Va-t-on bien sortir de la boucle au bout d'un certain temps ? Les valeurs obtenues au final sont-elles celles attendues ?

<h3> invariant de boucle </h3>

Un invariant de boucle est une **propriété attachée à la boucle, qui est vraie initialement, avant de commencer à exécuter la boucle, et maintenue vraie pour toute itération de la boucle** (d'où son nom). Elle sera en particulier vraie à la sortie de la boucle.

Voici un algorithme de calcul avec un boucle conditionnelle et deux variables a et b ayant pour valeur un entier naturel.

On peut montrer que *"p = m x b" est un invariant de cette boucle.*

 - Avant le premier passage, p = 0 et m = 0, donc on a bien p = m x b.

- Supposons que p = m x b avant un passage dans la boucle : 

Les nouvelles de m et p après une boucle sont m' = m + 1 et p' = p + b. 

On a m' x b = (m + 1) x b = m x b + b = p + b = p'.

- La propriété est donc vérifiée après passage dans la boucle.

**Cette propriété est bien un invariant de boucle**, donc elle est vraie à la sortie de la boucle. Or à la sortie de la boucle, on a m = a , donc p = a x b.
Cette boucle permet donc le calcul des produit a et b (stocké dans p).

<h3> Terminaison </h3>

Un algorithme ne doit toujours comporter qu'un nombre fini d'étapes. Dans le cas de boucles non conditionnelles (boucle for), le nombre d'étapes est déterminé, donc fini.
Dans le cas de boucles conditionnelles (boucle while), on utilise la notion de **variant** : on choisit un variant, c'est-à-dire une expression (souvent une variable) telle que la suite des valeurs de cette expression au cours des itérations prend, à un moment donné, une valeur satisfaisant la condition d'arrêt.

Considérons le code suivant :

In [1]:
a = 10
x = 0
while x**2 < a:
    x = x + 1

Si la valeur de a est négative ou nulle, il n'y a aucun passage dans la boucle.
Sinon, la suite des valeurs de la variable x est 0 , 1 , 2 ... , n, où n est la première valeur supérieure ou égale à la racine carrée de la valeur de a. En considérant le variant x, on peut donc conclure que le nombre de passages dans la boucle est bien fini.

<i> <b> Exercices d'application </b> </i>

1) Reprendre l'exemple du produit de deux nombres étudié plus haut et montrer la terminaison de la boucle.

<h4 style="color:blue;">m est incrémenté de 1 à chaque passage dans la boucle, donc il va prendre les valeurs 0,1,2 ... et finira par atteindre a après a passages. La boucle va donc se terminer.</h4>

2) On effectue la division euclidienne de a par b, ou a et b sont deux entiers strictement positifs. Il s'agit donc de déterminer deux entiers q et r tels que : 
a = bq + r avec 0 <= r < b, dont on donne l'algorithme ci-dessous :
    
 - Prouver que a = bq + r est un invariant de boucle.
 - Prouver la terminaison de l'algorithme.
 
<i> Vous pouvez utiliser <a href ="http://pythontutor.com/visualize.html#mode=display"> Pythontutor </a> pour visualiser les opérations </i>

In [ ]:
q = 0 
r = a
while r >= b:
    q = q + 1
    r = r - b

<h4 style="color:blue;">
Au départ, on a q = 0, donc bq + r = r et r = a 
    
donc on a bien a = bq + r
    
On suppose que a = bq + r est vérifié à l'entrée de la boucle.
    
q devient alors q'= q + 1 et r devient r' = r - b.
    
Donc bq'+r'=b(q+1)+r-b=bq+b+r-b=bq+r=a

La propriété est donc alors bien vérifiée à la sortie de la boucle.
a = bq + r est bien un invariant de boucle.

A chaque boucle r diminue de b. Il prend les valeurs, a, a-b, a-2b ... b étant un entier strictement positif, la suite des valeurs de r est strictement décroissante et r finira par devenir inférieur à b pour un nombre n de boucles. 
    
En effet a-nb est inférieur à b dès lors que n supérieur à (a-b)/b ou n supérieur à a/b-1.</h4>

<h3> c. Coût d'un algorithme </h3>

Quand on écrit un algorithme, on doit réfléchir pour qu'il soit le plus efficace possible. On doit notamment se poser la question de son temps d'exécution. Si on traite une liste de 10^7 valeurs, puis une liste de 10^8 valeurs, le temps de traitement sera-t-il multiplié par 10 ? Ce coût dépend de l'algorithme, de la liste, de la machine, et même du langage utilisé.

Pour comparer deux algorithmes, on cherche à déterminer l'ordre de grandeur de ce coût en fonction de la taille des données. On parle de **complexité d'un algorithme**.
Pour évaluer cette complexité, **on se place toujours dans le "pire des cas"**, c'est-à-dire qu'on compte le nombre maximum d'opérations.

Reprenons l'exemple de la multiplication, dont le code est donné ci-dessous. On y a ajouté un "timer" pour déterminer le **temps d'exécution** du programme.

In [11]:
import time
start = time.time() #on récupère le temps au début de l'exécution
a = 100000
b = 10
m = 0
p = 0
while m < a:
    m = m + 1
    p = p + b
end = time.time() # on récupère le temps à la fin de l'exécution
duree = end - start
print(duree)

0.023723840713500977


Tester ce programme pour différentes valeurs de a. Que remarquez-vous ?

Le temps de calcul est proportionnel à a : on dit alors que le coût est <b>LINEAIRE</b>. 

Le coût en temps est accessible également par le **nombre d'affectations** car c'est *ce qui coûte* en ressources pour la machine (les tests étant négligeables). On peut estimer son ordre de grandeur en insérant un compteur :

In [1]:
N = 0  # compteur d'affectations
a = 100000
b = 10
m = 0
p = 0
N = N + 4  # 4 affectations
while m < a:
    m = m + 1
    p = p + b
    N = N + 2  # 2 affectation par passage dans la boucle
print(N)

200004


On trouve que N = 2a + 4

La boucle est effectuée exactement a fois, et il y a 2 affectations par boucle, donc 2a affectations en tout et on initialise avec 4 affectations.

Le nombre d'opérations est donc encore proportionnel à a : on dit que le cout est <b>LINEAIRE</b>.

On parlera de coût linéaire dès lors que le nombre d'opérations est représenté par une fonction affine en fonction de la taille des données.

L'ordre de grandeur de la complexité s'écrit alors **O(a)**.

<i> <b> D'autres exemples </b> </i>

1) Ecrire un programme permettant de calculer la somme des entiers de 1 à n. Le tester pour différentes valeurs de n. 
Vérifier que son coût est linéaire en comptant les opérations effectuées.

In [28]:
def somme(n):
    s = 0
    for i in range(1 , n + 1):
        s = s + i
    return s

somme(50)

#1 affectation par boucle et une affectation d'initialisation donc a+1, donc linéaire.

1275

<p> 2) On considère le programme ci-dessous. Le tester pour différentes valeurs de n. Noter ces valeurs et le temps correspondant, puis utiliser un tableur pour visualiser le nuage de points (n, durée(n)). Que remarquez-vous ? On parle ici de coût quadratique (fonction du second degré). </p>
<p> Ajouter un compteur d'opération, et reprendre les questions précédentes pour avoir le nombre d'opérations en fonction de n.</p>

In [2]:
import time
start = time.time()
x = 0
n = 1000
nop = 2
    
for i in range(n):
    x = x + i
    nop = nop + 1
    for j in range(n):
        x = x + j
        nop = nop + 1
print (nop)
end = time.time()
duree = end - start
print(duree)


1001002
0.13147783279418945


<h4 style="color:blue;"> On trouve : (n, nop, duree)

    
1000 / 1001002 / 0,147 

2000 / 4002002 / 0,54

3000 / 9003002 / 1,39

4000 / 16004002 / 2,3

5000 / 25005002 / 3,90

10000 / 100010002 / 14,39 </h4>


<h2> 2. Parcours séquentiel </h2>

On va étudier ici quelques algorithmes nécessitant de parcourir une liste (on peut également parcourir un tuple). Le parcours séquentiel consiste à parcourir la liste élément par élément, en suivant l'ordre des éléments.

<h3> exemple 1 : calcul d'une moyenne </h3>

Ecrire une fonction permettant de calculer la moyenne d'une liste de nombres. Pensez à toutes les vérifications vues ci-dessus. Déterminer le coût de cette fonction.

In [7]:
def moyenne(t):
    s = 0
    for i in t:
        s = s + i
    m = s/len(t)
    return m

assert(moyenne([1, 5, 3])) == 3, "La moyenne n'est pas 3."
assert(moyenne([2.5, -5.5, 3])) == 0

<p style="color:blue;">Invariant de boucle ?
la boucle est une boucle for, donc bornée : l'algo se termine

<h3> exemple 2 : recherche d'une occurence </h3>

On cherche ici à rechercher de manière séquentielle la présence d'une valeur dans un tableau, qui peut être une liste, un tuple, ou une chaîne de caractères. 
On cherche une valeur ou un caractère précis, en le comparant à toutes les valeurs du tableau.
On donne ci-dessous deux scripts Python répondant à cette question.

In [15]:
import time
import random

def recherche(x , t):
    
    n = len(t)
    
    i = 0
    while i < n and x != t[i]:
        i = i + 1
    if i < n:
        return i

    
L = [i for i in range(1000000)]
random.shuffle(L)
start = time.time()
recherche(10, L)
end = time.time()
print(end-start)


1.9489080905914307


In [7]:
def recherche2(x ,t):
    for i in range(len(t)):
        if t[i] == x:
            return i

Etudier ces deux programmes, et déterminer leur coût.

1er programme : 2 affectation au départ, et au maximum n=len(t) boucles avec une affectation dans chaque, donc le coût est n+2. Il est linéaire.

2ème programme : le coût est nul ?

<h3> exemple 3 : recherche d'un extremum </h3>

On dispose d'un ensemble de nombres (liste, tuple) dont on cherche un extremum (maximum, minimum ou les deux).
L'algorithme est le suivant :
- si la liste est non vide, on suppose que le maximum est le premier élément
- on parcourt la liste et chaque fois qu'on rencontre un élément plus grand, on dit que c'est le nouveau maximum provisoire.

Ecrire une fonction **maxmin(liste)** qui prend une liste de nombres en paramètres et renvoie son maximum et son minimum. Quel est son coût ?

In [8]:
def maxmin(t):
    max = t[0]
    min = t[0]
    for valeur in t:
        if valeur > max:
            max = valeur
        if valeur < min:
            min = valeur
    return (max, min)

print(maxmin([-2 , 3 , 54]))

(54, -2)


S'inspirer de cet algorithme pour écrire une fonction **max(f, a, b)** qui permet de déterminer une valeur approchée du maximum d'une fonction f sur un intervalle [a,b].

In [1]:
def f(x):
    return x**2

def max(f,a,b):
    maxi = f(a)
    pas = (b-a)/100
    for i in range (101):
        if f(a + i * pas) > maxi:
            maxi = f(a + i * pas)
    return maxi

print(max(f, -15 ,2))
        

225


<h2> <b> 3. Exercices d'application </b> </h2>

<b> Exercice 1 </b> : 
On attribue un numéro à divers personnages. Ceci se traduit par la liste Pers = [['Julie' , 1] , ['Tom' , 2] , ['Sam' , 3] , ['Lea' , 4] , ['Charlie' , 5]].
Ecrire une fonction qui permet de rechercher de manière séquentielle si un personnage de numéro x (entier) quelconque est présent dans la liste. Cette fonction doit renvoyer le nom du personnage en question s'il est présent dans la liste, False sinon.

In [3]:
def rechercher(perso ,tab):
    assert type(perso) == list, "perso est une liste"
    
    occurence = False   # 1 affect.
    
    for i in range(len(tab)):  # balaye la liste : pour n éléments, on passe tout en revue
        if tab[i] == perso:
            occurence = True   # 1 affectation
            
    return occurence

Pers = [['Julie' , 1] , ['Tom' , 2] , ['Sam' , 3] , ['Lea' , 4] , ['Charlie' , 5]]
rechercher(['Tom', 2], Pers)

# liste de n éléments
# O(n) = n + 1 + 1 = n + 2  ~ n    LINEAIRE

True

<b> Exercice 2 </b> : On veut écrire une fonction verifie_tri qui prend en paramètres une liste d'entiers et qui renvoie True si la liste est correctement triée dans l'ordre croissant et False sinon.
Proposer des tests pour cette fonction, puis écrire son code.

In [7]:
def verifie_tri(liste):
    """ Renvoie True si la liste d'entier est triée"""
    for i in range(len(liste) - 1):
        if liste[i] > liste[i + 1]:
            return False
    return True
        
L = [ 1, 3, 2, 4, 5]
verifie_tri(L)

# liste de n éléments
# O(n) = n     LINEAIRE 
 

False

<b> Exercice 3 </b> : Ecrire une fonction permute qui prend en argument une liste de mots et modifie la liste en permutant le mot le plus court en nombre de lettres  avec le premier mot de la liste. La fonction ne renvoie rien.

In [4]:
def permute(liste):
    
    mot_min = liste[0]
    
    for i in range(1, len(liste)):
        if len(liste[i]) < len(mot_min):
            temp = liste[i]
            liste[i] = liste[0]
            liste[0] = temp
        
        
liste_noms = ['Julie' , 'Tom', 'Sam', 'Lea', 'Charlie' ]

permute(liste_noms)
liste_noms

# liste de n éléments
# 3 affectations pendant la permutation et une affectation au départ
# O(n) = 3n  + 1 ~ 3n ~ n    LINEAIRE 
 

['Lea', 'Julie', 'Tom', 'Sam', 'Charlie']

<b> Exercice 4 </b> : Ecrire un programme pour rechercher un mot dans un texte. Le mot et le texte sont des variables de type str. Etudier le coût de l'algorithme en fonction de la longueur du texte et de celle du mot.
    

In [6]:
def rech_mot(chaine, mot):
    """ Renvoie True si le mot est dans le texte chaine """
    liste = chaine.split(" ")
    if mot in liste:
        return True
    else:
        return False

texte = "je suis content d'être là mon ptit Bob !"
rech_mot(texte, "Bob")

# on balaye une fois la chaîne de n éléments
# O(n) = n     LINEAIRE 
 

['je', 'suis', 'content', "d'être", 'là', 'mon', 'ptit', 'Bob', '!']


True

In [14]:
def rech_mot2(chaine, mot):
    """ Renvoie True si le mot est dans le texte chaine """
    
    temp = ""   # definit un mot temporaire
    
    i = 0
    
    while i < len(chaine):   # balaye le chaine de caractère
        
        j = 0
        
        while j < len(mot):   # balaye le mot
            
            if chaine[i + j] == mot[j]:  # teste si la lettre de la chaine correspond à la lettre du mot
                j = j + 1
                temp = temp + mot[j - 1]   # ajoute au mot temporaire
            else:
                j = len(mot) + 1   # Sinon, on sort de la boucle
        
        if temp == mot: 
            return True    # si le mot est trouvé, on sort de la boucle
        i = i + 1
        
    return False        
            
texte = "je suis content d'être là mon ptit Bob !"
rech_mot2(texte, "jes")

# on balaye une fois la chaîne de n éléments, et pour chaque passage on balaye le mot recherché
# dans le pire des cas, le mot fait la longueur de la chaîne
# avec les affectations intermédiaires, on arrive à :
# O(n) = 2 + (n + 2) x (n + 2) = 2 + n² + 4n + 4 = n² + 4n + 6

# Donc O(n) ~ n²   QUADRATIQUE

False